Ratios
====
##OCF2 reads

![Image](https://github.com/pilarcormo/SNP_distribution_method/blob/master/arabidopsis_datasets/Centromere/OC_chr2_10kb/Rplot.ratios.png?raw=true)

________

![Image](https://github.com/pilarcormo/SNP_distribution_method/blob/master/arabidopsis_datasets/Centromere/OC_chr2_100kb/Rplot.ratios.png?raw=true)

First, I used 10 kb and 100 kb fragments to calculate the ratios.  The results are better when using 100 kb fragments, as the highest ratio is observed in the area where the actual causal mutation is located.  The decrease in the factor added to both numerator and denominator to calculate the ratio seems to have an effect when going from 1 to 0.1, while 0.01 looks really similar to the previous one. Although the trend of the values is the same, the mutation appears with higher clarity when this value is lower than 1. I am afraid that this decrease can also cause the lost of the causative mutation. 

##BCF2 reads

![Image](https://github.com/pilarcormo/SNP_distribution_method/blob/master/arabidopsis_datasets/Centromere/BCF2_chr3_10kb/Rplot.ratios.png?raw=true)

________

![Image](https://github.com/pilarcormo/SNP_distribution_method/blob/master/arabidopsis_datasets/Centromere/BCF2_chr3_100kb/Rplot.ratios.png?raw=true)

In this case, we cannot distinguish a point of the chromosome where the ratio is higher but this was expected as all the candidate SNPs they defined where scattered through all the beginning of the chromosome and there is not a clear linkage between hm SNPs. However, we are able to narrow the window to look for candidate SNPs. Again, when using 100kb fragments, the important ratios are highlighted.  

##sup#1 reads


#####Chromosome 4

![Image](https://github.com/pilarcormo/SNP_distribution_method/blob/master/Reads/Aw_sup1-2/Variant_calling/sup1_2_4/Rplot.ratiios_comparison.png?raw=true)
