Results 03/03 
====
from [Scheneeberger 2009](http://1001genomes.org/software/shoremap.html)

#####Chromosome 1 hm SNP distribution
![Image](https://github.com/pilarcormo/SNP_distribution_method/blob/master/chromosome1/Rplot.hm1.png?raw=true)

#####Chromosome 2 hm SNP distribution
![Image](https://github.com/pilarcormo/SNP_distribution_method/blob/master/chromosome2/Rplot.hm2.png?raw=true)

#####Chromosome 3 hm SNP distribution
![Image](https://github.com/pilarcormo/SNP_distribution_method/blob/master/chromosome3/Rplot.hm3.png?raw=true)

#####Chromosome 4 hm SNP distribution
![Image](https://github.com/pilarcormo/SNP_distribution_method/blob/master/chromosome4/Rplot.hm4.png?raw=true)

#####Chromosome 5 hm SNP distribution
![Image](https://github.com/pilarcormo/SNP_distribution_method/blob/master/chromosome5/Rplot.hm5.png?raw=true)

When using the chromosome 1 with SDM by creating a model genome, the densities are not high enough to provide a good approximation with SDM.

Results 05/03
====

from [BCF2 reads](http://1001genomes.org/software/shoremap.html)

In here I'm showing the overlapping plots with both densities. In red, the hm SNP density and in blue, the ht SNP density. By doing this, I can start guessing how the hm/ht ratio distribution will look like. 

#####Chromosome 1 
![Image](https://github.com/pilarcormo/SNP_distribution_method/blob/master/BCF2_chromosome1/Rplot.overlap_hmht.png?raw=true)

#####Chromosome 2 
![Image](https://github.com/pilarcormo/SNP_distribution_method/blob/master/BCF2_chromosome2/Rplot.overlap_hmht.png?raw=true)

#####Chromosome 3
![Image](https://github.com/pilarcormo/SNP_distribution_method/blob/master/BCF2_chromosome3/Rplot.overlap_hmht.png?raw=true)
#####Chromosome 4
![Image](https://github.com/pilarcormo/SNP_distribution_method/blob/master/BCF2_chromosome4/Rplot.overlap_hmht.png?raw=true)
#####Chromosome 5 
![Image](https://github.com/pilarcormo/SNP_distribution_method/blob/master/BCF2_chromosome5/Rplot.overlap_hmht.png?raw=true)
