Results 20/02
====

Instead of plotting the whole genome, I am focusing in the groups of contigs that have relevant ratios. In other words, I am plotting only the part of the genome where the causative SNP is located. With the 5 replicates used for the following genomes, the results obtained were similar to those. In general, a 10% of the genome remains after removing the non-important contigs. 

####From chr1_A 
![Image](https://github.com/pilarcormo/SNP_distribution_method/blob/master/arabidopsis_datasets/chr1_A_1/Perm_1902.rb/Plot_hyp.png?raw=true)

- The length of the group of contigs that form the peak of the distribution is 4282410 bp
- Location of causal mutation in correctly ordered genome: 14973755
- Candidate SNP position in permutation: 15130010

####From chr1_C (candidate SNP moved to the right)

![Image](https://github.com/pilarcormo/SNP_distribution_method/blob/master/arabidopsis_datasets/chr1_C_1/Perm_1902.2.rb/Plot_hyp.png?raw=true)

- The length of the group of contigs that form the peak of the distribution is 3486483 bp
- Location of causal mutation in correctly ordered genome: 16711444
- Candidate SNP position in permutation: 16739790

####From chr1_E (candidate SNP moved to the left)
![Image](https://github.com/pilarcormo/SNP_distribution_method/blob/master/arabidopsis_datasets/chr1_E_1/Perm_1902.2/Plot_hyp.png?raw=true)

- The length of the group of contigs that form the peak of the distribution is 3797783 bp
- Location of causal mutation in correctly ordered genome: 13617256
- Candidate SNP position in permutation: 13617256
